srtcore/netinet_any.h

```diff
@@ -32,28 +32,157 @@ struct sockaddr_any
+        void* object_begin = this;
+        memset(object_begin, 0, sizeof *this);
+
+        // Overriding family as required in the parameters
+        // and the size then accordingly.
+        sa.sa_family = domain == AF_INET || domain == AF_INET6 ? domain : AF_UNSPEC;
         len = size();
     }
 
-    socklen_t size() const
+    sockaddr_any(const sockaddr_storage& stor)
+    {
+        // Here the length isn't passed, so just rely on family.
+        set((const sockaddr*)&stor);
+    }
+
+    sockaddr_any(const sockaddr* source, socklen_t namelen = 0)
+    {
+        if (namelen == 0)
+            set(source);
+        else
+            set(source, namelen);
+    }
+
+    void set(const sockaddr* source)
+    {
+        // Less safe version, simply trust the caller that the
+        // memory at 'source' is also large enough to contain
+        // all data required for particular family.
+        if (source->sa_family == AF_INET)
+        {
+            memcpy(&sin, source, sizeof sin);
+            len = sizeof sin;
+        }
+        else if (source->sa_family == AF_INET6)
+        {
+            memcpy(&sin6, source, sizeof sin6);
+            len = sizeof sin6;
+        }
+        else
+        {
+            // Error fallback: no other families than IP are regarded.
+            sa.sa_family = AF_UNSPEC;
+            len = 0;
+        }
+    }
+
+    void set(const sockaddr* source, socklen_t namelen)
+    {
+        // It's not safe to copy it directly, so check.
+        if (source->sa_family == AF_INET && namelen >= sizeof sin)
+        {
+            memcpy(&sin, source, sizeof sin);
+            len = sizeof sin;
+        }
+        else if (source->sa_family == AF_INET6 && namelen >= sizeof sin6)
+        {
+            // Note: this isn't too safe, may crash for stupid values
+            // of source->sa_family or any other data
+            // in the source structure, so make sure it's correct first.
+            memcpy(&sin6, source, sizeof sin6);
+            len = sizeof sin6;
+        }
+        else
+        {
+            // Error fallback: no other families than IP are regarded.
+            void* object_begin = this;
+
+            // This embraces setting sockaddr::sa_family = AF_UNSPEC
+            // and sockaddr_any::len = 0.
+            memset(object_begin, 0, sizeof *this);
+        }
+    }
+
+    sockaddr_any(const in_addr& i4_adr, uint16_t port)
+    {
+        // Some cases require separately IPv4 address passed as in_addr,
+        // so port is given separately.
+        sa.sa_family = AF_INET;
+        sin.sin_addr = i4_adr;
+        sin.sin_port = htons(port);
+        len = sizeof sin;
+    }
+
+    sockaddr_any(const in6_addr& i6_adr, uint16_t port)
+    {
+        sa.sa_family = AF_INET6;
+        sin6.sin6_addr = i6_adr;
+        sin6.sin6_port = htons(port);
+        len = sizeof sin6;
+    }
+
+    static socklen_t size(int family)
+    {
+        switch (family)
```

Comment:
Please clean up this switch(). Also it would probably be a cleaner to have a single return value from this function with switch() setting the return value: that would fit common C/C++ patterns.